Relax mutation to aten.to in export

yushangdi opened the pull request:
Summary:
Today, we force aten.to to always copy under the hood and ban the mutation on the output of aten.to to address silent behavior divergence from eager mode. But in some cases, this is too strict because we only want to ban mutation if the output has an alias to it.

Here we relax it to: you can mutate the result of aten.to, but you cannot use it for any computation further.

You can:
```
    def forward(self, x):
        y = x + 1
        z = y.to("cpu")
        z.add_(5)
        return x
```

You can't:

```
    def forward(self, x):
        y = x + 1
        z = y.to("cpu")
        z.add_(5)
        return z
```


Test Plan:
```
buck2 run 'fbcode//mode/dev-nosan' fbcode//caffe2/test:test_export
buck2 run 'fbcode//mode/dev-nosan' fbcode//caffe2/test:test_export  -- -r  test_device_to_mutation
```

Differential Revision: D63672868

tugsbayasgalan commented on test/export/test_export.py:
```diff
@@ -3744,17 +3744,91 @@ def forward(self, x):
+        #         Module(), (torch.tensor(1, device="cpu"),)
+        #     )
+
+    def test_device_to_mutation_return_float(self):
```
Can we have a one test that forces copy? I think you can initialize int tensor and call float() on it.

yushangdi replied:
> Can we have a one test that forces copy? I think you can initialize int tensor and call float() on it.

i think the `test_device_to_mutation_not_aliasing` test should forces copy? We call `z.to(torch.int64)` on a float.

tugsbayasgalan replied:
Ah gotcha

tugsbayasgalan commented on test/export/test_export.py:
```diff
@@ -3744,17 +3744,91 @@ def forward(self, x):
+                y.add_(1)
+                return y, x
+
+        torch.export.export_for_training(Module(), (torch.tensor(1, device="cpu"),))
```
Today there is some problem with how this to logic is preserved in training IR. Could you also test this for torch.export.export?

yushangdi replied:
> Today there is some problem with how this to logic is preserved in training IR. Could you also test this for torch.export.export?

Not sure if this is a comment on the outdated version. I just exported to gh again. All tests should have both export_for_training and export.export now.

tugsbayasgalan replied:
Ok sounds good :)

tugsbayasgalan commented on torch/_subclasses/functional_tensor.py:
```diff
@@ -533,6 +546,20 @@ def unwrap(x):
                     )
                 else:
+                    # Check args before the operation
+                    for arg in args_unwrapped:
```
I think we discussed this before but i forgot hahah, how can this work when we don't have a proper functionlization rule for aten._to_copy that always aliases.

yushangdi replied:
> I think we discussed this before but i forgot hahah, how can this work when we don't have a proper functionlization rule for aten._to_copy that always aliases.

again, not sure if this is a comment on the outdated PR. I think we have a proper functionlization rule for aten._to_copy  now? register_aten_to_copy_py_functionalize_impl

yushangdi commented on torch/_functorch/_aot_autograd/collect_metadata_analysis.py:
```diff
@@ -298,6 +299,11 @@ def inner(*flat_args):
                 out_tensor_alias_counts[curr_storage] += 1
+
+                if isinstance(o, FunctionalTensor):
+                    out_storage = StorageWeakRef(o.elem.untyped_storage())
```
> Should we check the input as well?

the input is checked. I think the gh is not synced with Diff. Let me export to gh again.

tugsbayasgalan commented on torch/_subclasses/functional_tensor.py:
```diff
@@ -537,23 +554,49 @@ def unwrap(x):
+                        and func == torch.ops.aten._to_copy.default
+                        and torch._C.DispatchKey.Functionalize in func.py_kernels
+                    ):
+                        outs_unwrapped = func.py_kernels[
```
cc: @bdhirsh

yushangdi committed:
Relax mutation to aten.to in export (#137195)

Summary:

Today, we force aten.to to always copy under the hood and ban the mutation on the output of aten.to to address silent behavior divergence from eager mode. But in some cases, this is too strict because we only want to ban mutation if the output has an alias to it.

Here we relax it to: you can mutate the result of aten.to, but you cannot use it for any computation further.

Under export, aten.to() should always return a FunctionalTensor that aliases its input storage

If a tensor X with the above storage (S) is mutated, we need to flip a bit so that we can error IF:
- Any aliases of X are used in any operation (or if they are returned as graph outputs)
- If X is not a graph input, but its storage is part of a graph input





You can:
```
    def forward(self, x):
        y = x + 1
        z = y.to("cpu")
        z.add_(5)
        return x
```


You can't:

```
    def forward(self, x):
        y = x + 1
        z = y.to("cpu")
        z.add_(5)
        return z
```

Test Plan:
```
buck2 run 'fbcode//mode/dev-nosan' fbcode//caffe2/test:test_export
buck2 run 'fbcode//mode/dev-nosan' fbcode//caffe2/test:test_export  -- -r  test_device_to
buck2 run 'fbcode//mode/dev-nosan' //caffe2/test/functorch:test_aotdispatch -- -r test_aot_export_ban_dropout_mut_pre_dispatch
buck2 run 'fbcode//mode/dev-nosan' fbcode//caffe2/test:torch -- -r test_to$
```

Differential Revision: D63672868
